Bring Meson build definition more in line with CMake

dcbaker opened the pull request:
Update the Meson build definition to use Meson best practices, and bring into alignment with the CMake build files.

Particularly, This changes the meson build to only expose a single target, like the CMake does, controlled by an option with the same name. It additionally exposes the `GlobalUDLs` and `ImplicitConversions` options, which have behavioral implementations a consumer may wish to control. I have not exposed other options as they don't seem to be important for behavior.

The options names do not have the `JSON_` prefix, and Meson already has (sub)project option namespaces, so adding `JSON_` would worse (`-Dnlohmann_json:ImplicitConversions` vs `-Dnlohmann_json:JSON_ImplicitConversions`). 

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [ ]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header files `single_include/nlohmann/json.hpp` and `single_include/nlohmann/json_fwd.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

dcbaker committed:
meson: Indent code inside an if block

for better readability

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker committed:
meson: set a minimum Meson version

Without a version set meson will give no developer warnings, including
deprecations. 0.64 was selected as it's quite old, it's the newest
version supported by muon (a pure C Meson implementation), and there's
nothing complicated going on here.

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker committed:
meson: use `override_dependency()` to set dependencies

This simplifies the use of json as a subproject.

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker committed:
meson: use `install_subdir` for headers

This makes a single call to install the entire directory, and doesn't
need an update if any new headers are added. It also will simplify
bringing the Meson and CMake builds into allignment on how they handle
the multi-header vs single-header setups.

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker committed:
meson: handle single header and multiheader the same way CMake does

This uses a meson option (set in `meson_options.txt`) to control whether
multi-header or single-header setup is wanted.

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker committed:
meson: set the C++ standard to C++11

Matching the CMake as closely as possible, as Meson doesn't have C++11
feature checks like CMAke does.

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker committed:
meson: add support for the GlobalUDLs option

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker committed:
meson: add support for the ImplictConversions option

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

gregmarr commented on meson.build:
```diff
@@ -1,24 +1,43 @@
+# SPDX-License-Identifier: <<spdxid>>
+# Copyright © 2024 Dylan Baker
```
@dcbaker Was this addition intentional?

dcbaker replied:
No, sigh. My editor is trying to be helpful at least.

nlohmann commented:
Is meson_options.txt really needed? If so, can it be moved into a subdirectory?

dcbaker commented:
Meson options must be specified in a `meson_options.txt` (or for fairly new versions of Meson in a `meson.options` file), and that file has to be in the same directory as the root `meson.build` file. Meson options are read-only from the `meson.build` files, and are read in before the parsing of the `meson.build` file starts.

dcbaker commented:
I know that the meson_options.txt seems to be a blocker here, if I opened a new PR with the patches up until meson_options.txt is added, would that allow those changes to move faster?

nlohmann commented:
How is this related to #3885?

dcbaker commented:
It's not at all. I can comment there.

nlohmann commented:
Please do - I don't know Meson and will not be able to solve any of these issues.

heitbaum commented:
This PR improves the meson build significantly. All the headers are now being installed. It doesn’t address the .cmake files (but that is a seperate issue.)

it would be good to see this merged.

https://github.com/nlohmann/json/issues/3885#issuecomment-1356724948

```
./usr/lib/pkgconfig/nlohmann_json.pc
./usr/include/nlohmann/detail/string_concat.hpp
./usr/include/nlohmann/detail/string_escape.hpp
./usr/include/nlohmann/detail/value_t.hpp
./usr/include/nlohmann/detail/macro_scope.hpp
./usr/include/nlohmann/detail/abi_macros.hpp
./usr/include/nlohmann/detail/iterators/iter_impl.hpp
./usr/include/nlohmann/detail/iterators/primitive_iterator.hpp
./usr/include/nlohmann/detail/iterators/json_reverse_iterator.hpp
./usr/include/nlohmann/detail/iterators/iteration_proxy.hpp
./usr/include/nlohmann/detail/iterators/iterator_traits.hpp
./usr/include/nlohmann/detail/iterators/internal_iterator.hpp
./usr/include/nlohmann/detail/hash.hpp
./usr/include/nlohmann/detail/input/lexer.hpp
./usr/include/nlohmann/detail/input/json_sax.hpp
./usr/include/nlohmann/detail/input/input_adapters.hpp
./usr/include/nlohmann/detail/input/position_t.hpp
./usr/include/nlohmann/detail/input/binary_reader.hpp
./usr/include/nlohmann/detail/input/parser.hpp
./usr/include/nlohmann/detail/exceptions.hpp
./usr/include/nlohmann/detail/json_custom_base_class.hpp
./usr/include/nlohmann/detail/conversions/to_json.hpp
./usr/include/nlohmann/detail/conversions/from_json.hpp
./usr/include/nlohmann/detail/conversions/to_chars.hpp
./usr/include/nlohmann/detail/json_ref.hpp
./usr/include/nlohmann/detail/output/serializer.hpp
./usr/include/nlohmann/detail/output/binary_writer.hpp
./usr/include/nlohmann/detail/output/output_adapters.hpp
./usr/include/nlohmann/detail/json_pointer.hpp
./usr/include/nlohmann/detail/meta/is_sax.hpp
./usr/include/nlohmann/detail/meta/type_traits.hpp
./usr/include/nlohmann/detail/meta/identity_tag.hpp
./usr/include/nlohmann/detail/meta/detected.hpp
./usr/include/nlohmann/detail/meta/cpp_future.hpp
./usr/include/nlohmann/detail/meta/void_t.hpp
./usr/include/nlohmann/detail/meta/call_std/end.hpp
./usr/include/nlohmann/detail/meta/call_std/begin.hpp
./usr/include/nlohmann/detail/meta/std_fs.hpp
./usr/include/nlohmann/detail/macro_unscope.hpp
./usr/include/nlohmann/adl_serializer.hpp
./usr/include/nlohmann/json_fwd.hpp
./usr/include/nlohmann/thirdparty/hedley/hedley_undef.hpp
./usr/include/nlohmann/thirdparty/hedley/hedley.hpp
./usr/include/nlohmann/ordered_map.hpp
./usr/include/nlohmann/json.hpp
./usr/include/nlohmann/byte_container_with_subtype.hpp
```

dcbaker commented:
I've rebased on the current develop branch to fix conflicts and added my signed-off-by

nlohmann commented on meson_options.txt:
```diff
@@ -0,0 +1,18 @@
+    'MultipleHeaders',
+    type: 'boolean',
+    value: true,
+    description: 'Use non-amalgomated version of the library',
```
```suggestion
    description: 'Use non-amalgamated version of the library',
```

nlohmann commented on meson_options.txt:
```diff
@@ -0,0 +1,18 @@
+    type: 'boolean',
+    value: true,
+    description: 'Enable implicit conversions',
+)
```
There are more flags, see https://json.nlohmann.me/integration/cmake/#cmake-options for example. Though not are equally interesting, I wonder why here only theres three have been selected.

dcbaker replied:
It's been a while, I remember for sure that the `MultipleHeaders` option is essential to get the same behavior out of Meson and CMake, especially when using the Meson definitions in an embedded project. I don't remember why I added the other two.

dcbaker replied:
@nlohmann: I looked more, and I'm pretty sure that the reason that I added those two options in particular is that the related defines are exposed publicly in the pkg-config file. So for Meson and CMake to generate equivalent .pc files we'd need that.

Now that I'm looking at the generated CMake Target, it looks like JSON_DISABLE_ENUM_SERIALIZATION, JSON_DIAGNOSTICS, JSON_DIAGNOSTIC_POSITIONS, and JSON_USE_LEGACY_DISCARDED_VALUE_COMPARISON are exposed in CMake, so maybe we should update the pkg-config CMake generates to expose those as well?

nlohmann replied:
Sounds plausible, but I would not know how :)

nlohmann reviewed (changes requested):
Please also document `meson.build` and `meson_options.txt` in `FILES.md`.

dcbaker committed:
Add meson information to FILES.md

Signed-off-by: Dylan Baker <dylan@pnwbakers.com>

dcbaker commented:
I had forgotten about this, oops.

This still needs more work, but I've rebased on the devel branch and added some Meson information to the FILES.md file.

I want to look at the pkg config file thing some more as there is some discrepency between the .pc file and the cmake export files.
